### PR TITLE
Deploy to HuggingFace Template

### PR DESCRIPTION
**Pull request by freddyaboulton**
Add some instructions and scripts to deploy the todos app to huggingface spaces. 

Persistent storage is a paid feature on hugging face, this template will automatically backup the sqlite files a to a huggingface dataset.

### CODE REVIEW

File: todos2-hf/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+python-fasthtml>=0.0.8
+uvicorn>=0.29
+python-multipart
+huggingface-hub>=0.20.0
+sqlite_utils
```
Comment (jph00) [inline todos2-hf/requirements.txt:5]:
```suggestion

```

This isn't required -- sqlite_minutils is included in fasthtml.

**Review by jph00 — CHANGES_REQUESTED** [review]
Thank you for this - great stuff! BTW best to at-notify me of new PRs/issues so I don't miss them.

Is the dockerfile needed? In the past I've used hf spaces without needing a dockerfile, which was a little simpler. We'd like to use whatever the simplest approach is that works! (And if that's having a dockerfile, that's fine -- just checking...)

File: todos2-hf/requirements.txt (hunk 1 above)
Comment (freddyaboulton) [reply todos2-hf/requirements.txt]: Removed!

**Comment by freddyaboulton** [comment]
@jph00 removed sqlite-utils from the requirements file. About the Dockerfile - we could deploy this app using the gradio space template and that would remove the need for the Dockerfile. I thought that'd be confusing though since fasthtml is not gradio. I think down the line it would be nice if there was a built-in fasthtml template for spaces and that would require using the Dockerfile. 

I'll leave it up to you - happy to implement whichever one :)

**Comment by jph00** [comment]
Can you explain the dockerfile thing a bit more Freddy? Does it currently
work without the dockerfile? If so, what makes you think it might be
confusing?

A builtin template sounds very interesting. What might that do/contain?
Would it make your example simpler/easier?

**Comment by freddyaboulton** [comment]
The Dockerfile is only needed for deploying to spaces. Locally, you can run the `main.py` and it will work as normal. 

All Hugging Face spaces are built from Dockerfiles. It's just that for Gradio/Streamlit spaces, the Dockerfile is abstracted away from the end user. We can remove the Dockerfile from this PR, but then we'd have to deploy FastHTML apps as Gradio applications (reference [here](https://huggingface.co/docs/hub/spaces-sdks-gradio#create-a-new-gradio-space)). Nothing wrong with that but it felt a bit weird since FastHTML is its own framework.

By built-in template, I meant one of the pre-configured framework options available from the space creation screen. Like JupyterLab and Argilla:

image

The Dockerfile would still need to exist in this PR in that case. It would just make getting started with FastHTML on Spaces faster.

**Comment by jph00** [comment]
Makes perfect sense - thanks! I agree choosing 'gradio' would be confusing.

Having an SDK option for FastHTML would be super. Anything we can do to make that happen? Should we aim for that for day-1 release? (I'm guessing that'll be ~2 weeks away.)

**Pull request merged**